Query settings

dwijnand opened the pull request:
Builds on #2934.

WIP, just to share with @eed3si9n.

dwijnand committed:
Extract onExecCommand

dwijnand committed:
Pass State to NetworkChannel

dwijnand committed:
Add and handle GetSetting

eed3si9n commented on main/src/main/scala/sbt/internal/CommandExchange.scala:
```diff
@@ -74,7 +74,7 @@ private[sbt] final class CommandExchange {
         {
           s.log.info(s"new client connected from: ${socket.getPort}")
-          val channel = new NetworkChannel(newChannelName, socket)
+          val channel = new NetworkChannel(newChannelName, socket, s)
```
NetworkChannel represents a connection from some user. I don't think a state object should be owned by it since it's sort of misleading since the "state" there represents a past state.
I think a better approach is to pass in `(onQuery: Query => Unit)` here that it can call back, and handle query in the `CommandExchange`.

dwijnand replied:
In terms of "past state" I think using `onQuery: Query => Unit ` would amount to the same thing.

Imo, let's go with what works, and then use test cases to correct it.

eed3si9n replied:
It might work but I see it more of a breach of responsibility.

eed3si9n commented on main/src/main/scala/sbt/internal/server/NetworkChannel.scala:
```diff
@@ -71,15 +71,24 @@ final class NetworkChannel(val name: String, connection: Socket, state: State) e
   }
 }
+
+final case class GetSettingReq[A](settingKey: SettingKey[A]) extends CommandMessage
```
1. Probably move to protocol subproject. 
2. Rename to `SettingQuery`?
3. Avoid case class.

dwijnand replied:
Because SettingKey is defined in main-settings it's not trivial to move this to the protocol subproject.

dwijnand replied:
Also can't define this in the Contraband Schema Language: https://github.com/sbt/contraband/issues/74

eed3si9n commented on main/src/main/scala/sbt/internal/server/NetworkChannel.scala:
```diff
@@ -71,15 +71,24 @@ final class NetworkChannel(val name: String, connection: Socket, state: State) e
 }
+
+final case class GetSettingReq[A](settingKey: SettingKey[A]) extends CommandMessage
+final case class GetSettingRsp[A](value: Option[A]) extends EventMessage
```
1. Probably move to protocol subproject.
2. Rename to `SettingEvent`?
3. Avoid case class.

dwijnand replied:
Yeah, that's the WIP part. Not sure about "SettingEvent", could be confusing.

eed3si9n replied:
`SettingQueryResponse`?

dwijnand replied:
It breaks my natural vertical alignment, but sure.

eed3si9n reviewed (changes requested):
A few comments.

dwijnand committed:
Delete stale contraband classes

This is due to sbt/contraband#75.

dwijnand committed:
Delete SettingQuery instances we don't need

.. now that there's a copy in the git history

dwijnand commented:
Looks like this using `telnet`:

```
{"type":"SettingQuery","setting":"scalaVersion"}
{"type":"SettingQueryResponse","values":["2.12.1"]}
{"type":"SettingQuery","setting":"organization"}
{"type":"SettingQueryResponse","values":["com.dwijnand"]}
{"type":"SettingQuery","setting":"homepage"}
{"type":"SettingQueryResponse","values":["None"]}
```

eed3si9n commented:
Why do we get multiple values back? If it's because of project aggregation, we probably should remove that and force the client to pass in fully qualified settings.

dwijnand commented:
At least from aggregation. Maybe something else too. Delegation is also involved. I'm not sure if that's configuration delegation or project delegation.

Ultimately I was emulating show, except with structured dara rather than println toString.

By "fully qualified" do you mean it can't depend on the "current" project? What about depending on configuration delegation?

eed3si9n commented:
> By "fully qualified" do you mean it can't depend on the "current" project? 

No. The notion of current project shouldn't exist in querying as it's state dependent.

> What about depending on configuration delegation?

I think that's fine as long as it's stable. (It is right?)

dwijnand commented:
> > By "fully qualified" do you mean it can't depend on the "current" project?
> No. The notion of current project shouldn't exist in querying as it's state dependent.

OK, then that means creating alternative parsers for the string.

> > What about depending on configuration delegation?
> I think that's fine as long as it's stable. (It is right?)

It's not stable across reloads (or reboots) I assume, FWIW.

dwijnand committed:
Drop aggregation in querying settings

dwijnand committed:
Disallow SettingQuery relying on currentProject

Introduce a specialised scopedKeyParser on SettingQuery to redefine the
"projectRef" parser to never match "*" or omitted project refereneces.

dwijnand committed:
Introduce Def displayBuildRelative/showBuildRelativeKey

dwijnand committed:
Extract Act.taskKeyExtra

dwijnand committed:
Don't breach responsibility, NetworkChannel

Only depend on BuildStructure and currentBuild, not the whole of State!

dwijnand committed:
Bring back global project-axis reference

dwijnand commented:
@eed3si9n:

+ [x] `NetworkChannel` no longer owns a `State`
+ [x] data types moved to protocol project
+ [x] renamed to `SettingQuery`/`SettingQueryResponse`
+ [x] avoided case class, used CSL
+ [x] Removed use of aggregation
+ [x] Removed dependency on the current/session project, allows for explicit project/ThisBuild/Global

Could you give this a quick once over?

dwijnand commented:
Trailing concerns:

+ stale State (e.g change a file, reload on server, query settings returns old values)
+ proper structured JSON (i.e don't toString)

dwijnand commented:
I rebased on latest 1.0.x and remove the upgrade of the sbt-contraband plugin as it depends on a breaking bincompat change in sjson-new that was failing the tests, which I CBA to deal with right now.

> java.lang.NoClassDefFoundError: sjsonnew/LNil$

eed3si9n commented:
Merged!
